### PR TITLE
[Feature] Integrate MoonEP dispatcher for FSDP expert-parallel training

### PR DESCRIPTION
**Pull request by jayhenry**
## Summary

This PR adds an optional `dispatcher="moonep"` path for node-local BF16 MoE training. Native FSDP2 remains the sole owner of expert parameters, optimizer state, and checkpoint identity; MoonEP owns only the communication/VMM execution workspace.

- Add a model-scoped MoonEP runtime and adapt it to XTuner's six-stage dispatcher API.
- Land FSDP all-gathered BF16 expert weights directly into MoonEP VMM aliases; the private VMM workspace owns the physical expert layout.
- Dispatch activations and weights, reuse the existing grouped GEMM path (extended for dynamic expert gradients), and return duplicated expert gradients in BF16 before FSDP reduce-scatter, joined once per layer.
- Support MTP (shared and unshared weights), Domino intra-layer micro-batching, sequence parallelism, activation recompute, and `torch.compile`.
- Unify the EP dispatcher weight/routing contracts; keep routing counts on device via `torch.histc`.
- Integrate DCP/HF persistence, activation/router offload, optimizer swap, Muon, and explicit runtime teardown.
- Add a reproducible Qwen3.5 DeepEP/MoonEP acceptance gate (`tests/acceptance/`).
- Fix PyTorch 2.12 FSDP `_StridedShard` recognition in distributed grad-norm calculation.

```mermaid
flowchart LR
    A["FSDP all-gather: complete BF16 experts"] --> B["MoonEP VMM direct landing"]
    B --> C["Dispatch + grouped GEMM + combine"]
    C --> D["BF16 duplicated-gradient return"]
    D --> E["FSDP reduce-scatter"]
    E --> F["FP32 shard gradient + optimizer state"]
```

## Branch status

Rebased onto `main` and organized as three commits:

1. `[Fix]` route the grad-norm placement check through `RuntimeLayout.is_sharded_placement` so FSDP2 + EP `_StridedShard` bookkeeping placements (no longer a `Shard` subclass on PyTorch 2.12) contribute their all-reduce.
2. `[Feature]` the MoonEP integration.
3. `[Test]` the acceptance gate.

Design notes and the full per-run validation report are kept out of the PR and maintained separately.

## Design

MoonEP is imported lazily only when selected. The integration validates a versioned backend capability before allocating resources, so DeepEP, All2All, AGRS, and non-EP configurations do not acquire a MoonEP dependency.

The hot path is ordered with CUDA events and GPU-side EP barriers. The profiler regression gate verifies:

- no complete home-expert weight copy in the direct path;
- no complete local duplicated-gradient temporary;
- no CUDA device/event/stream host synchronization between planning and duplicated-gradient handoff.

FSDP post-all-gather hooks install the VMM landing tensors while preserving the original DTensor parameters. The backward path returns BF16 home gradients to those parameter edges; existing FSDP communication then reduce-scatters them and produces FP32 sharded gradients for the FP32 optimizer update.

## Supported scope

- PyTorch `2.12.1+cu132`.
- Single-node BF16 EP2/EP4/EP8; formal end-to-end acceptance uses FSDP2 + EP4 on 8 GPUs.
- TP1.
- MTP, Domino micro-batching, sequence parallelism, activation recompute, and compile.
- Synchronous/asynchronous DCP, HF export/load, AdamW, swap AdamW, and Muon.

Not claimed in this first version: Expert TP, FP8 experts, cross-node MoonEP, FSDP `no_sync`, pipeline parallelism, or decoding.

## External MoonEP dependency

This XTuner branch expects MoonEP XTuner integration API v3. The validated companion implementation is commit `d4494473fb0932dcc35f3a44a0e3b31827f5e282` on the MoonEP `xtuner-integration` branch.

**The companion MoonEP patch still needs to be published separately before this PR can be merged or reproduced outside the development environment.** The XTuner package does not add MoonEP as a mandatory dependency.

### Acceptance harness

`tests/acceptance/` (config-locked by `tests/engine/test_moonep_acceptance.py`, compared by `xtuner/_testing/moonep_acceptance.py`) runs Qwen3.5-35B-A3B on one node with 8 H200 GPUs, FSDP2 + EP4, BF16 parameters/reduction, Direct VMM landing, Triton grouped GEMM, and 20 training steps, changing only the dispatcher between the DeepEP and MoonEP legs. Default attention is FlexAttention + deterministic mode; `XTUNER_USE_FA3` and `XTUNER_DETERMINISTIC` are overridable.

### Validated behaviour (calibrated FlexAttention + deterministic gate)

On the functionally-equivalent pre-rebase implementation, MoonEP matches DeepEP on every loss and grad-norm curve (cosine similarity >= 0.99, mean relative difference < 1%) and reaches 100-112% of DeepEP steady-state throughput at MTP disabled and MTP1. GPU regression suites (MoonEP forward / MTP / Domino / SP / compile; DCP / HF / offload / optimizer / lifecycle; dispatcher and grouped-GEMM; FSDP2 + EP4 nested-shard grad norm on PyTorch 2.12 and 2.9) passed.

### Post-rebase verification on this branch

- Non-GPU regression: 40 passed — acceptance config-lock across all dispatcher/MTP/micro combinations, MoonEP dispatcher contract and six-stage seam, and device-only routing counts.
- GPU, FA3 (non-deterministic; FA3's Hopper backward has no deterministic path at head_dim 256), Qwen3.5-35B-A3B, FSDP2 + EP4, 20 steps, DeepEP vs MoonEP, MTP0 and MTP1, at 32k (micro1) and 16k (micro2) sequence length:
  - Numerical parity preserved: every curve vs DeepEP has cosine >= 0.99997, mean relative difference < 1.2%.
  - Throughput MoonEP/DeepEP: ~0.95 (micro1) / ~0.84 (micro2) — bit-for-bit consistent with both the pre-rebase branch and the 08-28 validated baseline under the same settings, i.e. no regression from the rebase. FA3 + non-deterministic is outside the calibrated gate's regime (which is FlexAttention + deterministic, where MoonEP is throughput-neutral) and this delta is present at the 08-28 baseline as well.

The formal FlexAttention + deterministic GPU gate has not yet been re-executed on the rebased branch; the acceptance harness is unchanged and runs it directly.

## Result

The first-version objective is met for single-node BF16 FSDP2 + EP training: MoonEP matches DeepEP numerically, preserves FSDP parameter/checkpoint ownership and a host-sync-free communication hot path, and is throughput-neutral in the calibrated acceptance configuration.

### CODE REVIEW

**Comment by jayhenry** [comment]
@claude review

File: xtuner/v1/module/dispatcher/moonep.py [hunk 1]
```diff
@@ -0,0 +1,751 @@
...
+        async_op: bool,
+    ) -> torch.Tensor:
+        ctx.invocation = invocation
+        ctx.save_for_backward(expert_output, route_weights)
+        assert invocation._plan is not None and invocation._runtime._buffer is not None
+        with torch.profiler.record_function("MoonEP::combine_forward"):
+            output, gathered_weights, done = invocation._runtime._buffer.combine(
+                plan=invocation._plan,
+                hidden_nvsh=expert_output,
+                hidden_scales_nvs=route_weights,
```
Comment (jayhenry) [inline xtuner/v1/module/dispatcher/moonep.py]: 不应该修改 moonep 增加 `hidden_scales_nvs` 参数。去掉它，然后将 hidden states 和 route weights 相乘放到 combine preprocess 中

File: xtuner/v1/model/moe/moe.py [hunk 2]
```diff
@@ -483,6 +513,11 @@ def forward(
             assert isinstance(loss_ctx, list) and len(loss_ctx) == len(seq_ctx), (
                 "seq_ctx_list and loss_ctx_list must be lists of the same length"
             )
+            if self._moonep_runtime is not None and len(seq_ctx) > self.config.intra_layer_micro_batch:
```
Comment (jayhenry) [inline xtuner/v1/model/moe/moe.py]: 改为通用检查 len(seq_ctx) == self.config.intra_layer_micro_batch

File: xtuner/v1/module/dispatcher/moonep.py [hunk 3]
```diff
@@ -0,0 +1,751 @@
...
+        ep_group: dist.ProcessGroup,
+        hidden_size: int,
+        intermediate_size: int,
+        num_experts: int,
+        top_k: int,
+        intra_layer_micro_batch: int,
+        staging_reference: bool,
+        num_sms: int = 64,
+    ) -> None:
+        self._backend = require_moonep_backend()
```
Comment (jayhenry) [inline xtuner/v1/module/dispatcher/moonep.py]: 直接检查是否安装 moonep，不做过度检查。
另外，在文件顶层 try import moonep 即可，后续引用也直接用 moonep，去掉 self._backend

File: xtuner/v1/module/router/greedy.py [hunk 4]
```diff
@@ -87,14 +87,14 @@ def forward(self, logits: torch.Tensor, rollout_routed_experts: torch.Tensor | N
 
         # moe forward
         # (e, )
-        tokens_per_expert = torch.histc(topk_ids, bins=self.n_routed_experts, min=0, max=self.n_routed_experts)
+        tokens_per_expert = torch.bincount(topk_ids.flatten(), minlength=self.n_routed_experts)
```
Comment (jayhenry) [inline xtuner/v1/module/router/greedy.py]: torch.bincount 存在强制 Host-Device 同步，改回原来的 torch.histc

File: xtuner/v1/module/decoder_layer/moe_decoder_layer.py [hunk 5]
```diff
@@ -297,14 +324,21 @@ def __init__(
...
         ep_tp_group = ep_tp_mesh._flatten().get_group() if ep_tp_mesh is not None else None
+        # EP membership is static for the layer. Keep the decision outside the
+        # compiled forward; Naive EP=1 execution retains its synchronous API.
+        self._async_combine = process_group is not None and process_group.size() > 1
         self.dispatcher = build_dispatcher(
             dispatcher=dispatcher,
             n_routed_experts=n_routed_experts,
             ep_group=process_group,
             tp_group=tp_group,
             ep_tp_group=ep_tp_group,
```
Comment (jayhenry) [inline xtuner/v1/module/decoder_layer/moe_decoder_layer.py:341]: 先hardcode bf16，去掉这个参数，assert 不支持 fp8

File: xtuner/v1/module/router/noaux_router.py [hunk 6]
```diff
@@ -133,20 +133,14 @@ def forward(self, logits, rollout_routed_experts: torch.Tensor | None = None) ->
...
         topk_weight = topk_weight * self.router_scaling_factor  # must multiply the scaling factor
 
-        # TODO: (yehaochen) `Dispatcher` calculate the distribution duplicatedly
-        tokens_per_expert = torch.histc(
-            topk_ids.float(),
-            bins=self.n_routed_experts,
-            min=0,
-            max=self.n_routed_experts,
-        )  # .view(self.ep_mesh.size(), -1)
+        tokens_per_expert = torch.bincount(topk_ids.flatten(), minlength=self.n_routed_experts)
```
Comment (jayhenry) [inline xtuner/v1/module/router/noaux_router.py]: 同样要使用不host sync的方式，比如 histc

File: xtuner/v1/ops/moe/cuda/group_gemm_cutlass.py [hunk 7]
```diff
@@ -3,25 +3,33 @@
 https://github.com/fanshiqing/grouped_gemm/blob/v1.1.4/grouped_gemm/ops.py
 Support torch compile."""
 
+import grouped_gemm_backend as backend
 import torch
 from torch import Tensor
 
 
-try:
```
Comment (jayhenry) [inline xtuner/v1/ops/moe/cuda/group_gemm_cutlass.py:10]: 保留 try except 的import 方式

File: xtuner/v1/module/dispatcher/fsdp_vmm_landing.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    """FSDP must not resize/free runtime-owned, non-resizable VMM storage."""
+    del fsdp_param
+
+
+def install_fsdp_vmm_landing(
+    *,
+    fsdp_root: nn.Module,
+    targets: Sequence[tuple[str, tuple[nn.Module, nn.Module], tuple[torch.Tensor, torch.Tensor]]],
+) -> tuple[FSDPParam, ...]:
+    """Bind routed expert FSDPParams to their two-generation VMM landings.
```
Comment (jayhenry) [inline xtuner/v1/module/dispatcher/fsdp_vmm_landing.py]: 将各种检查封装到一个私有函数，保证顶层的线性步骤

File: xtuner/v1/module/dispatcher/moonep.py [hunk 9]
```diff
@@ -0,0 +1,751 @@
...
+        )
+
+        self._buffer: Any | None = None
+        self._workspace: Any | None = None
+        self._fsdp_params: tuple[Any, ...] = ()
+        self._layers: list[tuple[str, tuple[nn.Module, nn.Module]]] = []
+        self._fixed_tokens_per_rank: int | None = None
+        self._closed = False
+
+    def bind_dispatcher(
```
Comment (jayhenry) [inline xtuner/v1/module/dispatcher/moonep.py]: 改名为 `build_dispatcher` 是否更合适？

File: xtuner/v1/module/dispatcher/moonep.py [hunk 10]
```diff
@@ -0,0 +1,751 @@
...
+
+class MoonEPPostCombineResult(TypedDict):
+    hidden_states: torch.Tensor
+
+
+class _MoonEPInvocation:
+    """Own one fresh MoonEP plan and all of its device-side completion
+    edges."""
+
+    def __init__(self, runtime: MoonEPRuntime, *, layer_id: int, grad_slot: int) -> None:
```
Comment (jayhenry) [inline xtuner/v1/module/dispatcher/moonep.py]: 将 runtime 整个传参是否参数过宽？

File: xtuner/v1/module/dispatcher/moonep.py [hunk 11]
```diff
@@ -0,0 +1,751 @@
...
+        tokens_per_expert: torch.Tensor,
+        topk_weights: torch.Tensor,
+        async_op: bool,
+    ) -> MoonEPDispatchResult:
+        hidden_nvsh, weights_nvs, cu_seqlens = _DispatchAutograd.apply(
+            hidden_states,
+            topk_ids,
+            tokens_per_expert,
+            topk_weights,
+            self,
```
Comment (jayhenry) [inline xtuner/v1/module/dispatcher/moonep.py]: 将 invocation 整个传参是否参数过宽？

File: xtuner/v1/module/dispatcher/moonep.py [hunk 12]
```diff
@@ -0,0 +1,751 @@
...
+        workspace = runtime._workspace
+        buffer = runtime._buffer
+        assert workspace is not None and buffer is not None and self._plan is not None
+
+        # FSDP has exposed its BF16 unsharded Parameters at this point.
+        # Staging changes only how their values reach the MoonEP workspace;
+        # completed home gradients are handed back before FSDP post-backward.
+        with torch.profiler.record_function("MoonEP::prepare_experts"):
+            home_parameters = self._current_home_parameters()
+            local_weights, gradient_targets, weights_ready = workspace.materialize(
```
Comment (jayhenry) [inline xtuner/v1/module/dispatcher/moonep.py]: 这个函数会 prefetch weight 产生GPU通信，如果想要 domino ep 正常工作，是不是应该将它放到 dispatcher.dispatch 函数中？

File: xtuner/v1/module/dispatcher/moonep.py [hunk 13]
```diff
@@ -0,0 +1,751 @@
...
+        pre_dispatched: MoonEPPreDispatchResult,
+        topk_weights: torch.Tensor,
+        async_op: bool = False,
+        decoding: bool = False,
+    ) -> MoonEPDispatchResult:
+        if decoding:
+            raise NotImplementedError("MoonEP fixed-S training dispatch does not implement decoding")
+        grad_slot = self._next_gradient_slot
+        self._next_gradient_slot = (grad_slot + 1) % self._runtime._intra_layer_micro_batch
+        return _MoonEPInvocation(self._runtime, layer_id=self._layer_id, grad_slot=grad_slot).dispatch(
```
Comment (jayhenry) [inline xtuner/v1/module/dispatcher/moonep.py]: 现在 dispatcher太浅，只做简单的转发。将 Invocation 退化成一个对象类，并没有 dispatch, combine 等方法，让 dispatcher 本身有具体的dispatch等逻辑，这样改是否合理？
